Make opensslconf.h include depending on target archs

shigeki opened the pull request:
This is a proposal to change `deps/openssl/conf/opensslconf.h` before upgrading 1.0.2a. 

@bnoordhuis @indutny I would like to have your feedback.

The current `deps/openssl/conf/opensslconf.h` is very hard to maintain with increasing supporting OS and CPU because it is generated by Configure script with specifying a target argument, where it includes sevral defines that depend on OS and architecture platform. See
https://github.com/shigeki/io.js/blob/old_upgrade_openssl102a/deps/openssl/doc/openssl_conf.pdf
is a table at a glance.

A new approach here is that we statically mapped `--dest-os` and `--dest-cpu` options in configure to the target of Configure in OpenSSL. The opensslconf.h in each target was created in advance by typing `deps/openssl/openssl/Configure {target}` and copied into `deps/openssl/conf/archs/{target}/opensslconf.h`.  After that we apply the fix of opensslconf.h according to iojs requirements.

We change `deps/openssl/conf/openssconf.h` so as to include each file according to its target by checking pre-defined compiler macros.

Here is a map table of configure options in iojs, target arch of Configure in OpenSSL and CI check.

| --dest-os | --dest-cpu | OpenSSL target arch  | CI  |
| --------- | ---------- | -------------------- | --- |
| linux     | ia32       | linux-elf            | o   |
| linux     | x32        | patched linux-x86\_\_64 | -   |
| linux     | x64        | linux-x86\_\_64         | o   |
| linux     | arm        | linux-armv4          | o   |
| linux     | arm64      | N/A                  | -   |
| mac       | ia32       | darwin-i386-cc       | o   |
| mac       | x64        | darwin64-x86-cc      | o   |
| win       | ia32       | VC-WIN32             | -   |
| win       | x64        | VC-WIN64A            | o   |
| solaris   | ia32       | solaris-x86-gcc      | o   |
| solaris   | x64        | solaris64-x86\_\_64-gcc | o   |
| freebsd   | ia32       | BSD-x86              | o   |
| freebsd   | x64        | BSD-x86\_\_64           | o   |
| openbsd   | ia32       | BSD-x86              | -   |
| openbsd   | x64        | BSD-x86\_\_64           | -   |
| others    | others     | linux-elf            | -   |

 --dest-os and --dest-cpu are mapped to pre-defined macros.

| --dest-os          | pre-defined macro         |
| ------------------ | ------------------------- |
| win                | \_\_WIN32                    |
| win(64bit)         | \_\_WIN64                    |
| mac                | \_\_\_\_APPLE\_\_\_\_ && \_\_\_\_MACH\_\_\_\_     |
| solaris            | \_\_\_\_sun                     |
| freebsd            | \_\_\_\_FreeBSD\_\_\_\_               |
| openbsd            | \_\_\_\_OpenBSD\_\_\_\_               |
| linux (not andorid)| \_\_\_\_linux\_\_\_\_ && !\_\_\_\_ANDROID\_\_\_\_ |
| android            | \_\_\_\_ANDROID\_\_\_\_               |

| --dest-cpu | pre-defined macro |
| ---------- | ----------------- |
| arm        | \_\_\_\_arm\_\_\_\_           |
| arm64      | \_\_\_\_aarch64\_\_\_\_       |
| ia32       | \_\_\_\_i386\_\_\_\_          |
| ia32(win)  | \_\_M\_\_IX86           |
| mips       | \_\_\_\_mips\_\_\_\_          |
| mipsel     | \_\_\_\_MIPSEL\_\_\_\_        |
| x32        | \_\_\_\_ILP32\_\_\_\_         |
| x64        | \_\_\_\_x86\_\_64\_\_\_\_        |
| x64(win)   | \_\_M\_\_X64            |

These are the list which is not implemented yet.

| --dest-os | --dest-cpu | OpenSSL target arch  | CI  |
| --------- | ---------- | -------------------- | --- |
| linux     | mips       | linux-mips32,linux-mips64,linux64-mips64? | --- |
| linux     | mipsel     | ?                    | --- |
| android   | ia32       | android-x86          | --- |
| android   | arm        | android-armv7        | --- |
| android   | mips       | android-mips         | --- |
| android   | mipsel     | ?                    | --- |

Supported target arch list in OpenSSL can be obtained by typing `deps/openssl/openssl/Configure LIST`.

The CI result is fine with this change as 
https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/464/

If this approach is not accepted, I will be  back work on it as in https://github.com/shigeki/io.js/commit/2d5ca2ab56e70edc2e8db53c9002a0713f2e2744 that was the same way as we did in the past.

bnoordhuis commented:
@shigeki Can you add a Makefile target that regenerates the config files?  Doesn't have to be the top-level Makefile, adding a new one in deps/openssl is fine too.

indutny commented:
@bnoordhuis +1

bnoordhuis commented on deps/openssl/config/opensslconf.h:
```diff
@@ -1,324 +1,121 @@
+#elif defined(_WIN32) && defined(_M_X64)
+  #include "./archs/VC-WIN64A/opensslconf.h"
+#elif (defined(__FreeBSD__) || defined(__OpenBSD__)) && defined(__i386__)
+  #include "./archs/BSD-x86/opensslconf.h"
```
To be clear, the uppercase directory name is generated by openssl's Configure script?

Style nit: can you write this as `#include` or `# include`?  This indent style isn't used anywhere else.

shigeki replied:
The directory name is the same as the one specified in Configure script. I will change the style to put the hash bangs at the head of line.

shigeki commented:
@bnoordhuis Adding Makefile is a good idea. I will try it.

shigeki commented:
I've just updated this to add Makefile that can re-generate files of opensslconf.h for all supported targets and apply fixes to linux-x32, darwin64-x86_64-cc and WIN. But `Configure` in the iojs repository adds some unnecessary files and overwrites  existing files so that I added  backup, restore and cleanup rules in Makefile. Styles are also fixed. PTAL.

bnoordhuis commented on deps/openssl/config/Makefile:
```diff
@@ -0,0 +1,63 @@
+PERL    = perl
+CONFIGURE = ./Configure
+CONFIGOPT = no-shared no-symlinks
+ARCS = BSD-x86 BSD-x86_64 VC-WIN32 VC-WIN64A darwin64-x86_64-cc darwin-i386-cc linux-armv4 linux-elf linux-x86_64 solaris-x86-gcc solaris64-x86_64-gcc
```
s/ARCS/ARCHS/ maybe?

shigeki replied:
Right. I will fix this.

bnoordhuis commented on deps/openssl/config/Makefile:
```diff
@@ -0,0 +1,63 @@
+BACKUP_EXT = iojsbackup
+
+PHONY = all clean backup restore
+.PHONY: $(PHONY)
```
Shouldn't ARCHS (ARCS) be PHONY as well?

bnoordhuis commented on deps/openssl/config/Makefile:
```diff
@@ -0,0 +1,63 @@
+	cd ../openssl; $(PERL) $(CONFIGURE) $(CONFIGOPT) linux-x86_64 > /dev/null;
+	$(SED) -e 's/#define OPENSSL_CPUID_OBJ//' -e 's/#define RC4_CHUNK unsigned long/#define RC4_CHUNK unsigned long long/' -e 's/#define SIXTY_FOUR_BIT_LONG/#undef SIXTY_FOUR_BIT_LONG/' -e 's/#undef SIXTY_FOUR_BI/#define SIXTY_FOUR_BIT/' ../openssl/crypto/opensslconf.h > ./archs/$@/opensslconf.h
+
+# The current openssl relase does not use RC4 asm since it explicitly
```
Typo: s/relase/release/

bnoordhuis commented on deps/openssl/config/Makefile:
```diff
@@ -0,0 +1,63 @@
+# specified as `$asm=~s/rc4\-[^:]+//;` in
+# https://github.com/openssl/openssl/blob/OpenSSL_1_0_1-stable/Configure#L584
+# But iojs has used RC4 asm on MacOS for long time. Fix type of RC4_INT
+# into `unsinged int` in opensslconf.h of darwin64-x86_64-cc to work on
```
Typo: s/unsinged/unsigned/

indutny commented:
Yay, looking good to me!

shigeki commented:
I have to make spell checks for my all commits.

bnoordhuis commented on deps/openssl/config/Makefile:
```diff
@@ -0,0 +1,63 @@
+	@rm ../openssl/tools/c_rehash.bak
+
+clean:
+	find archs -name opensslconf.h -exec rm "{}" \;
```
Because you're already using perl, would it be an idea to replace the sed one-liners with perl one-liners?  That prevents GNU sed vs. BSD sed compatibility issues.  You also don't have to escape so much because you can just do `perl -i -pe 's!foo/bar!quux/qunc!g'`.

bnoordhuis replied:
Oh, and would it be possible to wrap the long lines?

shigeki replied:
Okay, I will try it.

shigeki commented:
@bnoordhuis I've just update `Makefile` to use perl instead of sed and long lines got shorten to be less than 80.  At the previous update, I forgot to include the commit for installer, so added it. All files now were spell-checked. 

New CI of https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/474/ seems to be fine except new failures of smartos64 and win2008 but they are not related. It's great arm tests are getting better. Thanks.

bnoordhuis commented:
There's a small typo in the last commit log (s/archtectures/architectures/) but apart from that LGTM.  Nice work.

shigeki commented:
Thanks, Ben. I fixed the spell of the commit log with my shame.  Landed in 29a33014617902328b295f46c8417f321f57f67f , 7d14aa0222ef22ecabf010f9aeb094ddecf60a1f , 5b0e5755a07231ce358131feabfec69e80fdfcd6 and 8bc8bd4bc206d23fe8936608cdd5450cf037b586.
